commit/merkleroot/rmn/client.go

```diff
@@ -19,13 +41,595 @@ type Client interface {
+
+	// For each lane update request send an observation request to at most 'minObservers' number of rmn nodes.
+	for sourceChain, updateRequest := range updateRequestsPerChain {
+		requestedNodes[sourceChain] = mapset.NewSet[uint32]()
+
+		// At this point we assume at least minObservers RMN nodes are available for each source chain.
+		for nodeID := range updateRequest.rmnNodes.Iter() {
+			if requestedNodes[sourceChain].Cardinality() >= c.minObservers {
+				break
+			}
+
+			requestedNodes[sourceChain].Add(nodeID)
+			if _, exists := requestsPerNode[nodeID]; !exists {
+				requestsPerNode[nodeID] = make([]*rmnpb.FixedDestLaneUpdateRequest, 0)
+			}
+			requestsPerNode[nodeID] = append(requestsPerNode[nodeID], updateRequest.data)
+		}
+	}
+
+	requestIDs := c.sendObservationRequests(destChain, requestsPerNode)
+	signedObservations, err := c.listenForRmnObservationResponses(
+		ctx, destChain, requestIDs, updateRequestsPerChain, requestedNodes)
+	if err != nil {
+		return nil, fmt.Errorf("listen for rmn observation responses: %w", err)
+	}
+
+	// Sanity check that we got enough signed observations, should never happen, an error must have been received
+	// in the previous step if we didn't get enough observations for each source chain.
+	if err := c.ensureEnoughSignedObservations(signedObservations); err != nil {
+		return nil, fmt.Errorf("unexpected error, ensuring enough signed observations: %w", err)
+	}
+
+	return signedObservations, nil
+}
+
+func (c *client) sendObservationRequests(
+	destChain *rmnpb.LaneDest,
+	requestsPerNode map[uint32][]*rmnpb.FixedDestLaneUpdateRequest,
+) (requestIDs mapset.Set[uint64]) {
+	requestIDs = mapset.NewSet[uint64]()
+
+	for nodeID, requests := range requestsPerNode {
+		req := &rmnpb.Request{
+			RequestId: newRequestID(),
+			Request: &rmnpb.Request_ObservationRequest{
+				ObservationRequest: &rmnpb.ObservationRequest{
+					LaneDest:                    destChain,
+					FixedDestLaneUpdateRequests: requests,
+				},
+			},
+		}
+
+		c.lggr.Infow("sending observation request", "node", nodeID, "requestID", req.RequestId)
+		if err := c.marshalAndSend(req, nodeID); err != nil {
+			c.lggr.Errorw("failed to send observation request", "node_id", nodeID, "err", err)
+			continue
+		}
+		requestIDs.Add(req.RequestId)
+	}
+	return requestIDs
+}
+
+func (c *client) listenForRmnObservationResponses(
+	ctx context.Context,
+	destChain *rmnpb.LaneDest,
+	requestIDs mapset.Set[uint64],
+	lursPerChain map[uint64]updateRequestWithMeta,
+	requestedNodes map[uint64]mapset.Set[uint32],
+) ([]rmnSignedObservationWithMeta, error) {
+	c.lggr.Infow("Waiting for observation requests", "requestIDs", requestIDs.String())
+
+	respChan := c.rawRmnClient.Recv()
+
+	finishedRequestIDs := mapset.NewSet[uint64]()
+	rmnObservationResponses := make([]rmnSignedObservationWithMeta, 0)
+	merkleRootsCount := make(map[uint64]map[string]int) // sourceChain -> merkleRoot -> count
```

Comment:
nit: looks like we do merkle root bytes => string => bytes, maybe just keep it as `Bytes32` in this map?